python/lsst/qa/explorer/classifyObjects.py

```diff
@@ -0,0 +1,371 @@
+        with open(clfFilenameMorph, "rb") as f:
+            filters = pickle.load(f)
+
+        filtersInCat = set(cat.columns.get_level_values(0))
+        if filtersInCat != set(filters):
+            raise RuntimeError("Not all filters required in file")
+
+        # addMagnitdues will probably be removed later as it will hopefully be superseded by Tim
+        cat = addMagnitudes(cat, dataRef, filters)
+        cat = self.run(cat, filters)
+
+        cat = ParquetTable(dataFrame=cat)
+        dataRef.put(cat, "deepCoadd_sg")
+
+    def run(self, cat, filters):
+        """Add colour based classification, morphology only based classification and their associated flags.
+
+        Parameters
+        ----------
+        cat : 'Pandas data frame'
+        filters : 'list'
+            List of the filters for the camera used
+
+        Returns
+        -------
+        Data frame with the two classification and two flag columns added and a new index containing the
+        columns required by the classifier
+
+        Notes
+        -----
+        For the morphology only classifier the features used are psf - cmodel in r, i and z, m, b, seeing in
+        g, r, i, z and Y, flux/flux error in g, r, i, z and Y, psf - cmodel error in r, i and z
+        For the colour and morphology classifier the features used are g - r, r - i, i - z, z - Y (all using
+        cmodel), psf - cmodel in g, r, i, z and Y, m, b, flux / flux error in g, r, i, z and Y, seeing in g,
+        r, i, z and Y, psf - cmodel error in g, r, i, z and Y
+        The columns must be used in this order
+
+        Adds columns to the input table for morphology only classification, morph + color classification and
+        their associated flags, adds a new index to the file that contains the columns used by the
+        classifier.
+        """
+        # Add required columns
+        cat = addSeeing(cat, filters)
+        cat = addColors(cat, filters)
+        cat = addMB(cat, filters)
+        cat = addSNFlux(cat, filters)
+
+        # To Do: DM-14539 - Move this to somewhere else
+        # put this into /datasets/hsc then the butler will get it
+        # Add classifier_pickle to obs_base datasets, copy deepCoadd_skyMap.
+        clfFilenameMorph = "/home/sr525/starGalaxy/clfIsotonicMorph.pickle"
+        with open(clfFilenameMorph, "rb") as f:
+            filters = pickle.load(f)
+            clfMorph = pickle.load(f)
+            colsToUseMorph = pickle.load(f)
+
+        featuresMorph = np.zeros((len(colsToUseMorph), len(cat)))
+        for (i, col) in enumerate(colsToUseMorph):
+            featuresMorph[i] = cat["featuresSGSep"][col].data
+
+        # Need to take out the nans but remember which rows they are and assign nans to the probability and
+        # add a flag
+        featureSumsMorph = np.sum(featuresMorph, axis=0)
+        cleanIdsMorph = np.where((featureSumsMorph == featureSumsMorph))[0]
```

Comment:
Consider using `np.isfinite(featuresSumMorph)` for clarity, e.g. `np.nonzero(np.isfinite(featuresSumMorph)` or `np.where(np.isfinite(...))`

Is there a reasonable way to avoid nans in the input data?

Reply:
NaNs come from other steps of the processing and I put them in for the magnitudes of objects with negative flux that calib.getMagnitude can't deal with.